<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Will Landau, Dr. Jarad Niemi" />

<meta name="date" content="2015-01-01" />

<title>heterosis package methodology</title>




<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title"><code>heterosis</code> package methodology</h1>
<h4 class="author"><em>Will Landau, Dr. Jarad Niemi</em></h4>
<h4 class="date"><em>2015</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#purpose"><span class="toc-section-number">1</span> Purpose</a></li>
<li><a href="#the-hierarchical-model"><span class="toc-section-number">2</span> The hierarchical model</a><ul>
<li><a href="#interpretations-of-model-parameters"><span class="toc-section-number">2.1</span> Interpretations of model parameters</a></li>
<li><a href="#detecting-heterosis-genes"><span class="toc-section-number">2.2</span> Detecting heterosis genes</a></li>
</ul></li>
<li><a href="#fitting-the-model"><span class="toc-section-number">3</span> Fitting the model</a></li>
</ul>
</div>

<div id="purpose" class="section level1">
<h1><span class="header-section-number">1</span> Purpose</h1>
<p>The <code>heterosis</code> package is a tool for analyzing data from certain RNA sequencing experiments. These experiments concern the phenomenon of heterosis, or hybrid vigor, in situations where two genetic lines are inbred to produce a hybrid. An individual gene has</p>
<ul>
<li>high-parent heterosis if it is <em>more</em> expressed in the hybrid than in either parent.</li>
<li>low-parent heterosis if it is <em>less</em> expressed in the hybrid than in either parent.</li>
<li>mid-parent heterosis if its expression level in the hybrid is significantly different from its mean expression level in the parents.</li>
</ul>
<p>The package uses a hierarchical modeling strategy, executed with a slice-sampling-within-Gibbs Markov chain Monte Carlo algorithm, to calculate the posterior probability of each form of heterosis for each gene. These heterosis probabilities enable the user to detect heterosis genes and rank them according to the strength of heterosis.</p>
</div>
<div id="the-hierarchical-model" class="section level1">
<h1><span class="header-section-number">2</span> The hierarchical model</h1>
<p>Let <span class="math">\(y_{n, g}\)</span> be the fully preprocessed RNA-sequencing read count for sample/library <span class="math">\(n\)</span> and feature/gene <span class="math">\(g\)</span>. The full hierarchical model is as follows.</p>
<p><span class="math">\[ \begin{align*}
&amp;y_{n,g} \stackrel{\text{ind}}{\sim} \text{Poisson}(\exp({\varepsilon}_{n, g} + \eta(n, g))) \\
&amp; \qquad {\varepsilon}_{n, g} \stackrel{\text{ind}}{\sim} \text{N}(0, \rho_n^2 \gamma_g^2) \\
&amp; \qquad \qquad \rho_n^2 \stackrel{\text{ind}}{\sim} \text{Inv-Gamma}\left (\text{shape} = \frac{\nu_\rho}{2}, \ \text{scale} =  \frac{\nu_\rho \tau_\rho^2}{2} \right) \\
&amp; \qquad \qquad \qquad \nu_\rho \stackrel{\text{}}{\sim} \text{U}(0, d_\rho) \\
&amp; \qquad \qquad \qquad \tau_\rho^2 \stackrel{\text{}}{\sim} \text{Gamma}(\text{shape} = a_\rho, \text{rate} = b_\rho) \\
&amp; \qquad \qquad \gamma_g^2 \stackrel{\text{ind}}{\sim} \text{Inv-Gamma}\left (\text{shape} = \frac{\nu_\gamma}{2}, \ \text{scale} =  \frac{\nu_\gamma \tau_\gamma^2}{2} \right) \\
&amp; \qquad \qquad \qquad \nu_\gamma \stackrel{\text{}}{\sim} \text{U}(0, d) \\
&amp; \qquad \qquad \qquad \tau_\gamma^2 \stackrel{\text{}}{\sim} \text{Gamma}(\text{shape} = a_\gamma, \text{rate} = b_\gamma) \\
&amp; \qquad \phi_g \stackrel{\text{ind}}{\sim} \text{N}(\theta_\phi, \sigma_\phi^2 \xi_{\phi, g}) \\
&amp; \qquad \qquad \theta_\phi \stackrel{\text{}}{\sim} \text{N}(0, c_{\phi }^2) \\
&amp; \qquad \qquad \sigma_\phi \stackrel{\text{}}{\sim} \text{U}(0, s_{\phi}) \\
&amp; \qquad \qquad \xi_{\phi, g} \stackrel{\text{ind}}{\sim} p(\xi_{\phi, g} \ | \ k_\phi, r_\phi) \\
&amp; \qquad \alpha_g \stackrel{\text{ind}}{\sim} \text{N}( \theta_\alpha, \sigma_\alpha^2 \xi_{\alpha, g}) \\
&amp; \qquad \qquad \theta_\alpha \stackrel{\text{}}{\sim} \text{N}(0, c_{\alpha}^2) \\
&amp; \qquad \qquad \sigma_\alpha \stackrel{\text{}}{\sim} \text{U}(0, s_{\alpha}) \\
&amp; \qquad \qquad \xi_{\alpha, g} \stackrel{\text{ind}}{\sim} p(\xi_{\alpha, g} \ | \ k_\alpha, r_\alpha) \\
&amp; \qquad \delta_g \stackrel{\text{ind}}{\sim} \text{N}(\theta_\delta, \sigma_{\delta}^2 \xi_{\delta, g}) \\
&amp; \qquad \qquad \theta_\delta\stackrel{\text{}}{\sim} \text{N}(0, c_{\delta}^2) \\
&amp; \qquad \qquad \sigma_\delta \stackrel{\text{}}{\sim} \text{U}(0, s_{\delta}) \\
&amp; \qquad \qquad \xi_{\delta, g} \stackrel{\text{ind}}{\sim} p(\xi_{\delta, g} \ | \ k_\delta, r_\delta)
\end{align*} \]</span></p>
<p>where</p>
<p><span class="math">\[ \begin{align*}
\eta(n, g) = \begin{cases}
\phi_g - \alpha_g &amp; \text{ library $n$ is in treatment group 1 (parent 1)} \\
\phi_g + \delta_g &amp; \text{ library $n$ is in treatment group 2 (hybrid)} \\
\phi_g + \alpha_g &amp; \text{ library $n$ is in treatment group 3 (parent 2)} \\
\end{cases}
\end{align*} \]</span></p>
<div id="interpretations-of-model-parameters" class="section level2">
<h2><span class="header-section-number">2.1</span> Interpretations of model parameters</h2>
<dl>
<dt>
<span class="math">\(\rho_n\)</span>
</dt>
<dd>
log (base <span class="math">\(e\)</span>) normalization factor of sample/library <span class="math">\(n\)</span>, accounting for sequencing depth, etc.
</dd>
<dt>
<span class="math">\(\sigma_\rho\)</span>
</dt>
<dd>
prior standard deviation of the <span class="math">\(\rho_n\)</span>’s.
</dd>
<dt>
<span class="math">\(s_\rho\)</span>
</dt>
<dd>
a constant, usually set to an arbitrary high value.
</dd>
<dt>
<span class="math">\({\varepsilon}_{n, g}\)</span>
</dt>
<dd>
accounts for overdispersion relative to the Poisson likelihood.
</dd>
<dt>
<span class="math">\(\gamma_g\)</span>
</dt>
<dd>
the feature/gene-specific standard deviation of <span class="math">\(\{{\varepsilon}_{1, g}, \ldots, {\varepsilon}_{N, g}\}\)</span> (where <span class="math">\(N\)</span> is the number of samples/libraries). It serves the same purpose as a gene-specific negative binomial dispersion parameter.
</dd>
<dt>
<span class="math">\(\nu\)</span>
</dt>
<dd>
an inverse-scaled-chisquare degrees of freedom parameter. The higher it is, the more the <span class="math">\(\gamma_g\)</span> parameters “shrink” towards <span class="math">\(\tau^2\)</span>.
</dd>
<dt>
<span class="math">\(\tau\)</span>
</dt>
<dd>
close to the square root of the mean of the <span class="math">\(\gamma_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\phi_g\)</span>
</dt>
<dd>
for feature/gene <span class="math">\(g\)</span>, the mean of the log expression levels of the parents.
</dd>
<dt>
<span class="math">\(\theta_\phi\)</span>
</dt>
<dd>
prior mean of the <span class="math">\(\phi_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\sigma_\phi\)</span>
</dt>
<dd>
prior standard deviation of the <span class="math">\(\phi_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\alpha_g\)</span>
</dt>
<dd>
for feature/gene <span class="math">\(g\)</span>, half the difference in log expression levels of the parents.
</dd>
<dt>
<span class="math">\(\theta_\alpha\)</span>
</dt>
<dd>
prior mean of the <span class="math">\(\alpha_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\sigma_\alpha\)</span>
</dt>
<dd>
prior standard deviation of the <span class="math">\(\alpha_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\delta_g\)</span>
</dt>
<dd>
for feature/gene <span class="math">\(g\)</span>, the log expression level of hybrid minus the mean of the log expression levels of the parents.
</dd>
<dt>
<span class="math">\(\theta_\delta\)</span>
</dt>
<dd>
prior mean of the <span class="math">\(\delta_g\)</span>’s.
</dd>
<dt>
<span class="math">\(\sigma_\delta\)</span>
</dt>
<dd>
prior standard deviation of the <span class="math">\(\delta_g\)</span>’s.
</dd>
</dl>
</div>
<div id="detecting-heterosis-genes" class="section level2">
<h2><span class="header-section-number">2.2</span> Detecting heterosis genes</h2>
<p>The model immediately gives a posterior probability of heterosis for each gene.</p>
<dl>
<dt>
<span class="math">\(P(\delta_g &gt; |\alpha_g| \ | \ \text{data})\)</span>
</dt>
<dd>
Posterior probability of high-parent heterosis.
</dd>
<dt>
<span class="math">\(P(\delta_g &lt; -|\alpha_g| \ | \ \text{data})\)</span>
</dt>
<dd>
Posterior probability of low-parent heterosis.
</dd>
<dt>
<span class="math">\(P(|\delta_g| &gt; {\varepsilon}\ | \ \text{data},\  {\varepsilon})\)</span>
</dt>
<dd>
Posterior probability of mid-parent heterosis given a threshold <span class="math">\({\varepsilon}\)</span>.
</dd>
</dl>
<p>The <code>heterosis</code> package estimates these probabilities using a Markov chain Monte Carlo (MCMC) algorithm. With parameter samples <span class="math">\(\alpha_g^{(m)}\)</span> and <span class="math">\(\delta_g^{(m)}\)</span> (<span class="math">\(m = 1, \ldots, M\)</span>) from the MCMC,</p>
<p><span class="math">\[ \begin{align*}
P(\delta_g &gt; |\alpha_g| \ | \ \text{data}) &amp;\approx M {{}^{-1}}\sum_{m = 1}^M P(\delta_g^{(m)} &gt; |\alpha_g| \ | \ \text{data}) \\
P(\delta_g &lt; -|\alpha_g| \ | \ \text{data}) &amp;\approx M {{}^{-1}}\sum_{m = 1}^M P(\delta_g^{(m)} &lt; -|\alpha_g^{(m)}| \ | \ \text{data}) \\
P(|\delta_g| &gt; {\varepsilon}\ | \ \text{data},\  {\varepsilon}) &amp;\approx M {{}^{-1}}\sum_{m = 1}^M P(|\delta_g^{(m)}| &gt; {\varepsilon}\ | \ \text{data},\  {\varepsilon}) \\
\end{align*} \]</span></p>
</div>
</div>
<div id="fitting-the-model" class="section level1">
<h1><span class="header-section-number">3</span> Fitting the model</h1>
<p>To fit the model and estimate heterosis probabilities, the <code>heterosis</code> package uses a slice-sampling-within-Gibbs Markov chain Monte Carlo algorithm, which is fully derived and explained in the manuscript. The package returns to the user MCMC parameter samples, heterosis probabilities, and the log full posterior density up to a proportionality constant (a useful convergence diagnostic).</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
