<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Will Landau" />

<meta name="date" content="2015-01-01" />

<title>fbseq package hierarchical model</title>




<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title"><a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package hierarchical model</h1>
<h4 class="author"><em>Will Landau</em></h4>
<h4 class="date"><em>2015</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#the-hierarchical-model"><span class="toc-section-number">2</span> The hierarchical model</a><ul>
<li><a href="#signal"><span class="toc-section-number">2.1</span> Signal</a></li>
<li><a href="#noise"><span class="toc-section-number">2.2</span> Noise</a></li>
<li><a href="#model-summary"><span class="toc-section-number">2.3</span> Model summary</a></li>
</ul></li>
<li><a href="#inference"><span class="toc-section-number">3</span> Inference</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>The <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package fits a hierarchical model to RNA-sequencing (RNA-seq) count data in fully Bayesian fashion. Publication of the method is pending. For now, here is the model.</p>
</div>
<div id="the-hierarchical-model" class="section level1">
<h1><span class="header-section-number">2</span> The hierarchical model</h1>
<p>Let <span class="math">\(y_{n, g}\)</span> be the fully preprocessed RNA-sequencing read count for replicate (dat column) <span class="math">\(n\)</span> (<span class="math">\(n = 1, \ldots, N)\)</span> and gene (data row) <span class="math">\(g\)</span> (<span class="math">\(g = 1, \ldots, G)\)</span>. Let <span class="math">\(X\)</span> be the <span class="math">\(N \times L\)</span> design matrix for gene-specific model coefficient parameters <span class="math">\(\boldsymbol{\beta}_g = (\beta_{1, g}, \ \ldots, \ \beta_{L, g})\)</span>. Let <span class="math">\(\boldsymbol{X}_n\)</span> be the <span class="math">\(n\)</span>’th row of <span class="math">\(X\)</span>. Conditioned on the parameters <span class="math">\(\boldsymbol{\beta}_{ g}\)</span> and <span class="math">\({\varepsilon}_{n, g}\)</span>, the <span class="math">\(y_{n, g}\)</span>’s are treated as independent and Poisson-distributed in the likelihood.</p>
<p><span class="math">\[ \begin{align*}
y_{n,g} \ | \ \boldsymbol{\beta}_g, \ {\varepsilon}_{n, g} \stackrel{\text{ind}}{\sim} \text{Poisson} \left (\exp \left (h_n + {\varepsilon}_{n, g} + \boldsymbol{X}_n \boldsymbol{\beta}_g  \right ) \right )
\end{align*} \]</span></p>
<div id="signal" class="section level2">
<h2><span class="header-section-number">2.1</span> Signal</h2>
<p>The parameters of interest are the <span class="math">\(\beta_{\ell, g}\)</span>’s and their hyperparameters. Conditional on hyperparameters <span class="math">\(\theta_\ell\)</span>, <span class="math">\(\sigma_\ell^2\)</span>, and <span class="math">\(\xi_{\ell, g}\)</span>, the <span class="math">\(\beta_{\ell, g}\)</span>’s are independent with normal distributions.</p>
<p><span class="math">\[ \begin{align*}
\beta_{\ell, g} \ | \ \theta_\ell, \ \sigma_\ell^2, \ \xi_{\ell, g} \stackrel{\text{ind}}{\sim} \text{Normal}(\theta_\ell, \ \sigma_\ell^2 \xi_{\ell, g})
\end{align*} \]</span></p>
<p>The design matrix <span class="math">\(X\)</span> should be chosen to substantiate the conditional independence assumptions for the <span class="math">\(\beta_{\ell, g}\)</span>’s in the model.</p>
<p>The hierarchical means <span class="math">\(\theta_\ell\)</span> of <span class="math">\(\beta_{\ell, 1}, \beta_{\ell, G}\)</span> are given normal priors. Conditional on the initialization constants <span class="math">\(c_\ell^2\)</span>, the <span class="math">\(\theta_\ell\)</span>’s are assumed to be independent.</p>
<p><span class="math">\[ \begin{align*}
\theta_\ell \ | \ c_\ell^2 \stackrel{\text{ind}}{\sim} \text{Normal}(0, \ c_\ell^2)
\end{align*} \]</span></p>
<p>The <span class="math">\(c_\ell\)</span>’s are constants and should be large so that the priors on the <span class="math">\(\theta_\ell\)</span>’s are diffuse and thus less informative than otherwise.</p>
<p>The <span class="math">\(\sigma_\ell\)</span> parameters are assumed to be independent conditional on initialization constants <span class="math">\(s_\ell\)</span>.</p>
<p><span class="math">\[ \begin{align*}
\sigma_\ell \ | \ s_\ell \stackrel{\text{ind}}{\sim} \text{Uniform}(0, \ s_\ell^2)
\end{align*} \]</span></p>
<p>This prior is equivalent to a <span class="math">\(\sigma_\ell^{-1} \text{I}(\sigma_\ell &lt; s_\ell)\)</span> prior on <span class="math">\(\sigma_\ell^2\)</span>. The <span class="math">\(s_\ell\)</span> constants should be chosen to be large.</p>
<p>The <span class="math">\(\xi_{\ell, g}\)</span> parameters are auxiliary variables used to assign different marginal hierarchical distributions to the <span class="math">\(\beta_{\ell, g}\)</span>’s. Let the the <span class="math">\(\xi_{\ell, g}\)</span>’s be conditionally independent given constants <span class="math">\(k_\ell\)</span>, <span class="math">\(q_\ell\)</span>, and <span class="math">\(r_\ell\)</span>, and let</p>
<p><span class="math">\[ \begin{align*}
\xi_{\ell, g} \stackrel{\text{ind}}{\sim} p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ \ r_\ell)
\end{align*} \]</span></p>
<p>If <span class="math">\(p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ r_\ell) = I(\xi_{\ell, g} = 1)\)</span>, then <span class="math">\(\beta_{\ell, g}\)</span> have conditionally independent normal distributions (default setting of <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> ). If <span class="math">\(p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ r_\ell) = Exp(\text{rate} = k_\ell)\)</span>, then the <span class="math">\(\beta_{\ell, g}\)</span>’s are independent with Laplace distributions given <span class="math">\(\theta_\ell\)</span> and <span class="math">\(\sigma_\ell^2\)</span> (default: mean <span class="math">\(\theta_\ell\)</span> and variance <span class="math">\(\sigma_\ell^2\)</span>). If <span class="math">\(p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ r_\ell) = \text{Inverse-Gamma}(q_\ell, \ r_\ell)\)</span>, then the <span class="math">\(\beta_{\ell, g}\)</span>’s are independent with Student-<span class="math">\(t\)</span> distributions given <span class="math">\(\theta_\ell\)</span> and <span class="math">\(\sigma_\ell^2\)</span> (default: mean <span class="math">\(\theta_\ell\)</span> and variance <span class="math">\(\sigma_\ell^2\)</span>). If <span class="math">\(p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ r_\ell) = \text{Half-Cauchy}(0, 1)\)</span>, then the <span class="math">\(\beta_{\ell, g}\)</span>’s are independent with horseshoe distributions given <span class="math">\(\theta_\ell\)</span> and <span class="math">\(\sigma_\ell^2\)</span>. For each <span class="math">\(\ell\)</span> separately, the user can choose among normal, Laplace, Student-<span class="math">\(t\)</span>, and horseshoe priors on the <span class="math">\(\beta_{\ell, g}\)</span>’s.</p>
</div>
<div id="noise" class="section level2">
<h2><span class="header-section-number">2.2</span> Noise</h2>
<p>The <span class="math">\(h_n\)</span> terms are constants estimated from the data before the MCMC. These play the role of customary log-scale RNA-seq normalization factors, accounting for replicate-specific nuisance effects such as different sequencing depths.</p>
<p>The <span class="math">\({\varepsilon}_{n, g}\)</span> terms are noise that the Poisson distribution does not account for. Conditional on their variances <span class="math">\(\gamma_g\)</span>, the <span class="math">\({\varepsilon}_{n, g}\)</span> parameters are independent with normal distributions.</p>
<p><span class="math">\[ \begin{align*}
{\varepsilon}_{n,g} \ | \ \gamma_g \stackrel{\text{ind}}{\sim} \text{Normal}(0, \ \gamma_g)
\end{align*} \]</span></p>
<p>The <span class="math">\(\gamma_g\)</span>’s are analogous to the negative binomial dispersions in more traditional RNA-seq models from packages like <code>edgeR</code>. Conditioned on parameters <span class="math">\(\nu\)</span> and <span class="math">\(\tau\)</span>, they have independent inverse-gamma distributions.</p>
<p><span class="math">\[ \begin{align*}
\gamma_g \ | \ \nu, \ \tau \stackrel{\text{ind}}{\sim} \text{Inverse-Gamma} \left ( \frac{\nu}{2}, \ \frac{\nu \tau}{2} \right )
\end{align*} \]</span></p>
<p>We can interpret <span class="math">\(\nu\)</span> as the degree to which the <span class="math">\(\gamma_g\)</span>’s “shrink” towards <span class="math">\(\tau\)</span>. Given the initialization constant <span class="math">\(d\)</span>, <span class="math">\(\nu\)</span> has a uniform prior distribution</p>
<p><span class="math">\[ \begin{align*}
\nu \sim \text{Uniform}(0, \ d)
\end{align*} \]</span></p>
<p>And <span class="math">\(\tau\)</span> is the prior center of the <span class="math">\(\gamma_g\)</span>’s (between the prior mean and prior mode). Given initialization constants <span class="math">\(a\)</span> and <span class="math">\(b\)</span>, <span class="math">\(\tau\)</span> has a gamma prior.</p>
<p><span class="math">\[ \begin{align*}
\tau \sim \text{Gamma}(a, \ \text{rate} = b)
\end{align*} \]</span></p>
</div>
<div id="model-summary" class="section level2">
<h2><span class="header-section-number">2.3</span> Model summary</h2>
<p><span class="math">\[ \begin{align*}
&amp;y_{n,g} \ | \ \boldsymbol{\beta}_g, \ {\varepsilon}_{n, g} &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Poisson} \left (\exp \left (h_n + {\varepsilon}_{n, g} + \boldsymbol{X}_n \boldsymbol{\beta}_g \right ) \right ) \\
&amp;\qquad \beta_{\ell, g} \ | \ \theta_\ell, \ \sigma_\ell^2, \ \xi_{\ell, g} &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Normal}(\theta_\ell, \ \sigma_\ell^2 \xi_{\ell, g}) \\
&amp;\qquad \qquad \theta_\ell \ | \ c_\ell^2 &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Normal}(0, \ c_\ell^2) \\
&amp;\qquad \qquad \sigma_\ell \ | \ s_\ell &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Uniform}(0, \ s_\ell^2) \\
&amp;\qquad \qquad \xi_{\ell, g} &amp;&amp;\stackrel{\text{ind}}{\sim} p(\xi_{\ell, g} \ | \ k_\ell, \ q_\ell, \ r_\ell) \\
&amp;\qquad {\varepsilon}_{n,g} \ | \ \gamma_g &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Normal}(0, \ \gamma_g) \\
&amp;\qquad \qquad \gamma_g \ | \ \nu, \ \tau &amp;&amp;\stackrel{\text{ind}}{\sim} \text{Inverse-Gamma} \left ( \frac{\nu}{2}, \ \frac{\nu \tau}{2} \right ) \\
&amp;\qquad \qquad \qquad \nu &amp;&amp;\sim \text{Uniform}(0, \ d) \\
&amp;\qquad \qquad \qquad \tau &amp;&amp;\sim \text{Gamma}(a, \ \text{rate} = b)
\end{align*} \]</span></p>
</div>
</div>
<div id="inference" class="section level1">
<h1><span class="header-section-number">3</span> Inference</h1>
<p>The <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package estimates posterior probabilities that depend on logical conjunctions of linear combinations of the <span class="math">\(\beta_{\ell,g}\)</span> parameters. Possible examples are below.</p>
<p><span class="math">\[ \begin{align*}
&amp;P\left( \left . \beta_{1, g} &gt; 0 \  \right | \ \text{data} \right) \\
&amp;P\left( \left . \beta_{2, g} - \beta_{3, g} &gt; 1 \ \right | \ \text{data} \right) \\
&amp;P\left( \left . \beta_{1, g} - \beta_{3, g} &gt; 0 \text{  and  } \beta_{2, g} - \beta_{3, g} &gt; \sqrt{2} \text{  and  } \beta_{4, g} &gt; -\log(\pi) \ \right | \ \text{data} \right) \\
\end{align*} \]</span></p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
