<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Will Landau" />

<meta name="date" content="2015-01-01" />

<title>fbseq package tutorial</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title"><code>fbseq</code> package tutorial</h1>
<h4 class="author"><em>Will Landau</em></h4>
<h4 class="date"><em>2015</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#check-your-system."><span class="toc-section-number">2</span> Check your system.</a></li>
<li><a href="#install-fbseq."><span class="toc-section-number">3</span> Install <code>fbseq</code>.</a><ul>
<li><a href="#option-1-install-directly-from-github."><span class="toc-section-number">3.1</span> Option 1: install directly from GitHub.</a></li>
<li><a href="#option-2-install-from-the-source."><span class="toc-section-number">3.2</span> Option 2: install from the source.</a></li>
</ul></li>
<li><a href="#workflow-overview"><span class="toc-section-number">4</span> Workflow overview</a><ul>
<li><a href="#create-a-scenario-object."><span class="toc-section-number">4.1</span> Create a <code>Scenario</code> object.</a><ul>
<li><a href="#example-scenario-from-paschold"><span class="toc-section-number">4.1.1</span> Example scenario from <span class="citation">Paschold et al. (2012)</span></a></li>
</ul></li>
<li><a href="#optional-create-a-configs-object."><span class="toc-section-number">4.2</span> (Optional) create a <code>Configs</code> object.</a></li>
<li><a href="#optional-create-a-starts-object."><span class="toc-section-number">4.3</span> (Optional) create a <code>Starts</code> object.</a></li>
<li><a href="#create-a-chain-object."><span class="toc-section-number">4.4</span> Create a <code>Chain</code> object.</a></li>
<li><a href="#call-fbseq-on-the-chain-object."><span class="toc-section-number">4.5</span> Call <code>fbseq</code> on the <code>Chain</code> object.</a></li>
<li><a href="#inspect-the-output."><span class="toc-section-number">4.6</span> Inspect the output.</a></li>
</ul></li>
</ul>
</div>

<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(fbseq)
<span class="kw">library</span>(methods)</code></pre>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>The <code>fbseq</code> package fits a hierarchical model to RNA-seq data in (nearly) fully Bayesian fashion. Although it was originally designed to fit RNA-seq data, it can also be used to analyze count data with tens of thousands of response variables and only a handful of observations each.</p>
<p>Please read the model vignette first. An understanding of the underlying hierarchical model is important for understanding how to use this package</p>
</div>
<div id="check-your-system." class="section level1">
<h1><span class="header-section-number">2</span> Check your system.</h1>
<p>You need to have at least R <span class="math">\(\ge\)</span> 3.2.0, along with the R packages <code>coda</code>, <code>methods</code>, <code>MASS</code>, <code>MCMCpack</code>, <code>parallel</code>, <code>plyr</code>, <code>pracma</code>, and <code>knitr</code>, all are available through the <a href="https://cran.r-project.org/">Comprehensive R Archive Network (CRAN</a>. With those requirements met, you can install <code>fbseq</code>, load it in an R session, create input, and analyze output. To actually run the underlying Markov chain Monte Carlo (MCMC) procedure, however, you need access to a machine with a <a href="http://www.nvidia.com/object/cuda_home_new.html">CUDA-enabled GPU</a>, along with the <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code> package</a>. <code>fbseqCUDA</code> is the internal engine of <code>fbseq</code>, and it is implemented in CUDA to provide necessary acceleration for the MCMC. <code>fbseq</code> and <code>fbseqCUDA</code> are kept separate for convenience. For example, you can set up input with <code>fbseq</code> and without <code>fbseqCUDA</code> on a low-end laptop, run the main algorithm remotely with both <code>fbseq</code> and <code>fbseqCUDA</code> on a CUDA-enabled <a href="http://www.nvidia.com/object/gpu-cloud-computing-services.html">cloud computing enterprise</a> such as <a href="http://aws.amazon.com/ec2/instance-types/">Amazon Web Services</a>, and analyze the output locally with <code>fbseq</code> and without <code>fbseqCUDA</code>. See the <code>fbseqCUDA</code> installation vignette for more details.</p>
</div>
<div id="install-fbseq." class="section level1">
<h1><span class="header-section-number">3</span> Install <code>fbseq</code>.</h1>
<div id="option-1-install-directly-from-github." class="section level2">
<h2><span class="header-section-number">3.1</span> Option 1: install directly from GitHub.</h2>
<p>For this option, you need the <code>devtools</code> package, available from CRAN or GitHub. Just open R and run</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;wlandau/fbseq&quot;</span>)</code></pre>
<p>and then if you have CUDA,</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install_github</span>(<span class="st">&quot;wlandau/fbseqCUDA&quot;</span>)</code></pre>
</div>
<div id="option-2-install-from-the-source." class="section level2">
<h2><span class="header-section-number">3.2</span> Option 2: install from the source.</h2>
<p>Open a command line program such as Terminal in Mac/Linux and enter the following commands.</p>
<pre><code>git clone git@github.com:wlandau/fbseq.git
R CMD build fbseq
R CMD INSTALL fbseq_0.0.tar.gz</code></pre>
<p>Note: you may have to replace <code>fbseq_0.0.tar.gz</code> by the name of whatever tarball comes out of <code>R CMD build</code>.</p>
<p>You can install <code>fbseqCUDA</code> next, and the steps are analogous. However, you may have to modify the top of the <code>src/Makevars</code> file before running <code>R CMD build</code>. See the <code>fbseqCUDA</code> package vignette for more detailed installation instructions.</p>
</div>
</div>
<div id="workflow-overview" class="section level1">
<h1><span class="header-section-number">4</span> Workflow overview</h1>
<p>A typical workflow only requires a few lines of code. However, it requires an understanding of the S4 classes of this package. The steps are as follows.</p>
<div id="create-a-scenario-object." class="section level2">
<h2><span class="header-section-number">4.1</span> Create a <code>Scenario</code> object.</h2>
<p><code>Scenario</code> is an S4 class for storing the count data, design matrix, and instructions for calculating posterior probabilities of quantities of interest.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">new</span>(<span class="st">&quot;Scenario&quot;</span>))</code></pre>
<pre><code>## Formal class 'Scenario' [package &quot;fbseq&quot;] with 6 slots
##   ..@ bounds      : num(0) 
##   ..@ contrasts   : list()
##   ..@ counts      : num[0 , 0 ] 
##   ..@ design      : num[0 , 0 ] 
##   ..@ propositions: list()
##   ..@ supplement  : list()</code></pre>
<p>Type <code>help(&quot;Scenario-class&quot;)</code> for details about the slots. When you have these slots ready, use the <code>Scenario</code> function to create a <code>Scenario</code> object.</p>
<pre class="sourceCode r"><code class="sourceCode r">scenario =<span class="st"> </span><span class="kw">Scenario</span>(<span class="dt">counts =</span> counts, <span class="dt">design =</span> design, <span class="dt">contrasts =</span> contrasts, <span class="dt">bounds =</span> bounds, <span class="dt">propositions =</span> propositions, <span class="dt">supplement =</span> supplement)</code></pre>
<p>With the exception of <code>Chain</code> objects, arguments passed to <code>Scenario</code> need to be explicitly passed to their slots. For example,</p>
<pre class="sourceCode r"><code class="sourceCode r">scenario =<span class="st"> </span><span class="kw">Scenario</span>(counts, design, contrasts, bounds, propositions, supplement)</code></pre>
<p>would throw an error.</p>
<div id="example-scenario-from-paschold" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Example scenario from <span class="citation">Paschold et al. (2012)</span></h3>
<p>The package comes with a <code>Scenario</code> with data taken from <span class="citation">Paschold et al. (2012)</span>. The data is from an RNA-sequencing (RNA-seq) data experiment on maize described in the paper.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(paschold)
<span class="kw">head</span>(paschold@counts)</code></pre>
<pre><code>##                  B73_1 B73_2 B73_3 B73_4 Mo17_1 Mo17_2 Mo17_3 Mo17_4
## AC147602.5_FG004     1     0     0     2      0      1      0      0
## AC148152.3_FG001     3     4     6     0      8     17     18     20
## AC148152.3_FG005  2323  1533  1932  1945   2070   1582   2196   1882
## AC148152.3_FG006     2     1     0     2      4      4      5      3
## AC148152.3_FG008     3     3     4     1     31     40     45     49
## AC148167.6_FG001   672   598   728   713    743    655    821    824
##                  B73xMo17_1 B73xMo17_2 B73xMo17_3 B73xMo17_4 Mo17xB73_1
## AC147602.5_FG004          1          1          1          0          0
## AC148152.3_FG001          7          8         19         16          7
## AC148152.3_FG005       1532       1271       2168       2004       1305
## AC148152.3_FG006          4          1          1          0          4
## AC148152.3_FG008          8          9          4          3          2
## AC148167.6_FG001        728        731        809        764        754
##                  Mo17xB73_2 Mo17xB73_3 Mo17xB73_4
## AC147602.5_FG004          0          0          0
## AC148152.3_FG001         10         13         13
## AC148152.3_FG005       1454       1694       1362
## AC148152.3_FG006          1          1          2
## AC148152.3_FG008          3          6          2
## AC148167.6_FG001        609        801        769</code></pre>
<p>Here, the genes are rows (response variables) and the columns are RNA-seq libraries (observations, samples of genetic material). Each count is the relative measure of the expression of a gene in a laboratory sample of genetic material. The observations are divided into 4 groups:</p>
<ul>
<li><code>B73</code>: maize seedlings from an inbred corn population in Iowa.</li>
<li><code>Mo17</code>: seedlings from a different inbred corn population in Missouri.</li>
<li><code>B73xMo17</code>: hybrid seedlings produced from pollenating <code>B73</code> corn plants with <code>Mo17</code> pollen.</li>
<li><code>Mo17xB73</code>: hybrid seedlings produced from pollenating <code>Mo17</code> corn plants with <code>B73</code> pollen.</li>
</ul>
<p>The goal of this scenario is to detect which genes satisfy which of following conditions.</p>
<table>
<tbody>
<tr class="odd">
<td align="left">high-parent heterosis, both hybrids</td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left">all hybrid counts &gt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, both hybrids</td>
<td align="left"></td>
<td align="left">all hybrid counts &lt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis, <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><code>B73xMo17</code> counts &gt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><code>B73xMo17</code> counts &lt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis, <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><code>Mo17xB73</code> counts &gt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><code>Mo17xB73</code> counts &lt; <code>B73</code> and <code>Mo17</code> counts</td>
</tr>
</tbody>
</table>
<p>To translate this problem into a <code>Scenario</code>, we use the design matrix,</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@design</code></pre>
<pre><code>##            beta_1 beta_2 beta_3 beta_4 beta_5
## B73_1           1      1     -1      0      1
## B73_2           1      1     -1      0      1
## B73_3           1      1     -1      0     -1
## B73_4           1      1     -1      0     -1
## Mo17_1          1     -1      1      0      1
## Mo17_2          1     -1      1      0      1
## Mo17_3          1     -1      1      0     -1
## Mo17_4          1     -1      1      0     -1
## B73xMo17_1      1      1      1      1      1
## B73xMo17_2      1      1      1      1      1
## B73xMo17_3      1      1      1      1     -1
## B73xMo17_4      1      1      1      1     -1
## Mo17xB73_1      1      1      1     -1      1
## Mo17xB73_2      1      1      1     -1      1
## Mo17xB73_3      1      1      1     -1     -1
## Mo17xB73_4      1      1      1     -1     -1</code></pre>
<p>In words, the main effect parameters are <span class="math">\(\beta_{\ell, g}\)</span> for effect <span class="math">\(\ell = 1, \ldots, L = 5\)</span> and <span class="math">\(g = 1, \ldots, G = 39656\)</span>, and they describe linear combinations of the libraries within a gene. In the following table, a logarithmic scale is used for the interpretations.</p>
<table>
<tbody>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{1, g}\)</span></td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left">mean of <code>B73</code> and <code>Mo17</code></td>
</tr>
<tr class="even">
<td align="left"><span class="math">\(\beta_{2, g}\)</span></td>
<td align="left"></td>
<td align="left">half the difference between <code>Mo17</code> and the hybrids</td>
</tr>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{3, g}\)</span></td>
<td align="left"></td>
<td align="left">half the difference between <code>B73</code> and the hybrids</td>
</tr>
<tr class="even">
<td align="left"><span class="math">\(\beta_{4, g}\)</span></td>
<td align="left"></td>
<td align="left">difference between <code>B73xMo17</code> and <code>Mo17xB73</code></td>
</tr>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{5, g}\)</span></td>
<td align="left"></td>
<td align="left">experimental block effect</td>
</tr>
</tbody>
</table>
<p>With these interpretations, the table of conditions becomes</p>
<table>
<tbody>
<tr class="odd">
<td align="left">high-parent heterosis, both hybrids</td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left"><span class="math">\(\beta_{2, g} &gt; 0\)</span> and <span class="math">\(\beta_{3, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, both hybrids</td>
<td align="left"></td>
<td align="left"><span class="math">\(-\beta_{2, g} &gt; 0\)</span> and <span class="math">\(-\beta_{3, g} &gt; 0\)</span></td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis, <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(2\beta_{2, g} + \beta_{4, g} &gt; 0\)</span> and <span class="math">\(2\beta_{3, g} + \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(-2\beta_{2, g} - \beta_{4, g} &gt; 0\)</span> and <span class="math">\(-2\beta_{3, g} - \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis, <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(2\beta_{2, g} - \beta_{4, g} &gt; 0\)</span> and <span class="math">\(2\beta_{3, g} - \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis, <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(-2\beta_{2, g} + \beta_{4, g} &gt; 0\)</span> and <span class="math">\(-2\beta_{3, g} + \beta_{4, g} &gt; 0\)</span></td>
</tr>
</tbody>
</table>
<p>Each linear combination of the <span class="math">\(\beta_{\ell, g}\)</span>’s is a contrast, and each row of the table is a logical proposition. We want a posterior probability for each proposition. Hence, the Paschold scenario has the 12 contrasts,</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@contrasts</code></pre>
<pre><code>## $`high-parent_B73xMo17_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      1      0      0      0 
## 
## $`high-parent_B73xMo17_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      1      0      0 
## 
## $`low-parent_B73xMo17_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -1      0      0      0 
## 
## $`low-parent_B73xMo17_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -1      0      0 
## 
## $`high-parent_B73xMo17_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0      1      0 
## 
## $`high-parent_B73xMo17_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2      1      0 
## 
## $`low-parent_B73xMo17_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0     -1      0 
## 
## $`low-parent_B73xMo17_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2     -1      0 
## 
## $`high-parent_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0     -1      0 
## 
## $`high-parent_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2     -1      0 
## 
## $`low-parent_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0      1      0 
## 
## $`low-parent_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2      1      0</code></pre>
<p>the lower bounds in the inequalities (all zero),</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@bounds</code></pre>
<pre><code>## high-parent_B73xMo17_Mo17xB73_1 high-parent_B73xMo17_Mo17xB73_2 
##                               0                               0 
##  low-parent_B73xMo17_Mo17xB73_1  low-parent_B73xMo17_Mo17xB73_2 
##                               0                               0 
##          high-parent_B73xMo17_1          high-parent_B73xMo17_2 
##                               0                               0 
##           low-parent_B73xMo17_1           low-parent_B73xMo17_2 
##                               0                               0 
##          high-parent_Mo17xB73_1          high-parent_Mo17xB73_2 
##                               0                               0 
##           low-parent_Mo17xB73_1           low-parent_Mo17xB73_2 
##                               0                               0</code></pre>
<p>and the 6 logical propositions we care about, expressed in terms of the indices of the contrasts above.</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@propositions</code></pre>
<pre><code>## $`high-parent_B73xMo17_Mo17xB73`
## high-parent_B73xMo17_Mo17xB73_1 high-parent_B73xMo17_Mo17xB73_2 
##                               1                               2 
## 
## $`low-parent_B73xMo17_Mo17xB73`
## low-parent_B73xMo17_Mo17xB73_1 low-parent_B73xMo17_Mo17xB73_2 
##                              3                              4 
## 
## $`high-parent_B73xMo17`
## high-parent_B73xMo17_1 high-parent_B73xMo17_2 
##                      5                      6 
## 
## $`low-parent_B73xMo17`
## low-parent_B73xMo17_1 low-parent_B73xMo17_2 
##                     7                     8 
## 
## $`high-parent_Mo17xB73`
## high-parent_Mo17xB73_1 high-parent_Mo17xB73_2 
##                      9                     10 
## 
## $`low-parent_Mo17xB73`
## low-parent_Mo17xB73_1 low-parent_Mo17xB73_2 
##                    11                    12</code></pre>
<p>When the MCMC is run with the <code>fbseq</code> function, the package will estimate, for each gene, the probability that each of the 6 propositions is true.</p>
<p>Note: every <code>Scenario</code> object has an optional <code>suppelement</code> slot with additional information about the scenario. In this case, <code>paschold@supplement</code> is an empty list.</p>
</div>
</div>
<div id="optional-create-a-configs-object." class="section level2">
<h2><span class="header-section-number">4.2</span> (Optional) create a <code>Configs</code> object.</h2>
<p>The <code>Configs</code> S4 class contains the practical MCMC configuration parameters, such as the number of Monte Carlo iterations and the length of burnin, and important details about model specification. All of them have default settings, so for many simple applications, you should be able to get away with calling the <code>Configs</code> function with no arguments. For a full list of slots and their explanations, type <code>help(&quot;Configs-class&quot;)</code>.</p>
</div>
<div id="optional-create-a-starts-object." class="section level2">
<h2><span class="header-section-number">4.3</span> (Optional) create a <code>Starts</code> object.</h2>
<p>The <code>Starts</code> S4 class has the starting values of an MCMC chain. Usually, the user does not need to worry about it, as it is automatically generated for a given dataset and design with <code>generate_starts</code> inside the constructor of <code>Chain</code> class. However, the user may wish to modify individual slots. For more information, type <code>help(&quot;Starts-class&quot;)</code> and be sure to read the vignette on the hierarchical model.</p>
</div>
<div id="create-a-chain-object." class="section level2">
<h2><span class="header-section-number">4.4</span> Create a <code>Chain</code> object.</h2>
</div>
<div id="call-fbseq-on-the-chain-object." class="section level2">
<h2><span class="header-section-number">4.5</span> Call <code>fbseq</code> on the <code>Chain</code> object.</h2>
</div>
<div id="inspect-the-output." class="section level2">
<h2><span class="header-section-number">4.6</span> Inspect the output.</h2>
<div class="references">
<h1 id="references" class="unnumbered">References</h1>
<p>Paschold, Anja, Yi Jia, Caroline Marcon, Steve Lund, Nick B Larson, Cheng-Ting Yeh, Stephan Ossowski, et al. 2012. “Complementation Contributes to Transcriptome Complexity in Maize (<em>Zea Mays L.</em>) Hybrids Relative to Their Inbred Parents.” <em>Genome Research</em> 22 (12). Cold Spring Harbor Lab: 2445–54.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
