<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Will Landau" />

<meta name="date" content="2015-01-01" />

<title>fbseq package tutorial</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title"><a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package tutorial</h1>
<h4 class="author"><em>Will Landau</em></h4>
<h4 class="date"><em>2015</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#read-the-model-vignette-first."><span class="toc-section-number">2</span> Read the <a href="https://github.com/wlandau/fbseq/blob/master/vignettes/model.html">model vignette</a> first.</a></li>
<li><a href="#check-your-system."><span class="toc-section-number">3</span> Check your system.</a></li>
<li><a href="#install-fbseq"><span class="toc-section-number">4</span> Install <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a></a><ul>
<li><a href="#option-1-install-a-stable-release-recommended."><span class="toc-section-number">4.1</span> Option 1: install a stable release (recommended).</a></li>
<li><a href="#option-2-use-install_github-to-install-the-development-version."><span class="toc-section-number">4.2</span> Option 2: use <code>install_github</code> to install the development version.</a></li>
<li><a href="#option-3-build-the-development-version-from-the-source."><span class="toc-section-number">4.3</span> Option 3: build the development version from the source.</a></li>
</ul></li>
<li><a href="#install-an-mcmc-backend-package"><span class="toc-section-number">5</span> Install an MCMC backend package</a></li>
<li><a href="#quick-start"><span class="toc-section-number">6</span> Quick start</a></li>
<li><a href="#operational-details"><span class="toc-section-number">7</span> Operational details</a><ul>
<li><a href="#create-a-scenario-object."><span class="toc-section-number">7.1</span> Create a <code>Scenario</code> object.</a><ul>
<li><a href="#example-scenario-from-paschold"><span class="toc-section-number">7.1.1</span> Example scenario from <span class="citation">Paschold et al. (2012)</span></a></li>
<li><a href="#example-simulated-scenario-with-scenario_heterosis_model"><span class="toc-section-number">7.1.2</span> Example simulated scenario with <code>scenario_heterosis_model</code></a></li>
</ul></li>
<li><a href="#optional-create-a-configs-object."><span class="toc-section-number">7.2</span> (Optional) create a <code>Configs</code> object.</a></li>
<li><a href="#optional-create-a-starts-object."><span class="toc-section-number">7.3</span> (Optional) create a <code>Starts</code> object.</a></li>
<li><a href="#create-a-chain-object."><span class="toc-section-number">7.4</span> Create a <code>Chain</code> object.</a></li>
<li><a href="#call-fbseq-on-the-chain-object."><span class="toc-section-number">7.5</span> Call <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> on the <code>Chain</code> object.</a></li>
<li><a href="#choosing-a-backend-fbseqserial-or-fbseqcuda"><span class="toc-section-number">7.6</span> Choosing a backend: <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> or <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a></a></li>
<li><a href="#inspect-the-output."><span class="toc-section-number">7.7</span> Inspect the output.</a></li>
</ul></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>The <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package part of a collection of packages for the fully Bayesian analysis of RNA-sequencing (RNA-seq) count data, where a hierarchical model is fit with Markov chain Monte Carlo (MCMC). <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> is the user interface, and it contains top level functions for calling the MCMC and analyzing output. The other packages, <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> and <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> , are backend packages that run the MCMC behind the scenes. Only one is required. <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> can run on most machines, but it is slow for large datasets. <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> requires special hardware (i.e. a CUDA general-purpose graphics processing unit), but it’s much faster due to parallel computing.</p>
</div>
<div id="read-the-model-vignette-first." class="section level1">
<h1><span class="header-section-number">2</span> Read the <a href="https://github.com/wlandau/fbseq/blob/master/vignettes/model.html">model vignette</a> first.</h1>
<p>An understanding of the underlying hierarchical model is important for understanding how to use this package. For best viewing, download the html file to your desktop and then open it with a browser.</p>
</div>
<div id="check-your-system." class="section level1">
<h1><span class="header-section-number">3</span> Check your system.</h1>
<p>See the “Depends”, “Imports”, and “Suggests” fields of the “package’s <a href="https://github.com/wlandau/fbseq/blob/master/DESCRIPTION">DESCRIPTION</a> R version and R package requirements.</p>
</div>
<div id="install-fbseq" class="section level1">
<h1><span class="header-section-number">4</span> Install <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a></h1>
<div id="option-1-install-a-stable-release-recommended." class="section level2">
<h2><span class="header-section-number">4.1</span> Option 1: install a stable release (recommended).</h2>
<p>For <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> , you can navigate to a <a href="https://github.com/wlandau/fbseq/releases">list of stable releases</a> on the project’s <a href="https://github.com/wlandau/fbseq">GitHub page</a>. Download the desired <code>tar.gz</code> bundle, then install it either with <code>install.packages(..., repos = NULL, type=&quot;source&quot;)</code> from within R <code>R CMD INSTALL</code> from the Unix/Linux command line.</p>
</div>
<div id="option-2-use-install_github-to-install-the-development-version." class="section level2">
<h2><span class="header-section-number">4.2</span> Option 2: use <code>install_github</code> to install the development version.</h2>
<p>For this option, you need the <code>devtools</code> package, available from CRAN or GitHub. Open R and run</p>
<pre><code>library(devtools)
install_github(&quot;wlandau/fbseq&quot;)</code></pre>
</div>
<div id="option-3-build-the-development-version-from-the-source." class="section level2">
<h2><span class="header-section-number">4.3</span> Option 3: build the development version from the source.</h2>
<p>Open a command line program such as Terminal in Mac/Linux and enter the following commands.</p>
<pre><code>git clone git@github.com:wlandau/fbseq.git
R CMD build fbseq
R CMD INSTALL ...</code></pre>
<p>where <code>...</code> is replaced by the name of the tarball produced by <code>R CMD build</code>.</p>
</div>
</div>
<div id="install-an-mcmc-backend-package" class="section level1">
<h1><span class="header-section-number">5</span> Install an MCMC backend package</h1>
<p><a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> and <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> , are backend packages that run the MCMC behind the scenes. Only one is required. <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> can run on most machines, but it is slow for large datasets. <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> requires special hardware (i.e. a CUDA general-purpose graphics processing unit), but it’s much faster due to parallel computing. Installation is similar to that of <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> and is detailed in their respective <code>README.md</code> files and package vignettes.</p>
</div>
<div id="quick-start" class="section level1">
<h1><span class="header-section-number">6</span> Quick start</h1>
<p>After installing <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> and <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> , the following should take a couple seconds to run. The example walks through an example data analysis and shows a few key features of the package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(fbseq)

back_end =<span class="st"> &quot;serial&quot;</span> <span class="co"># change this to &quot;CUDA&quot; to use fbseqCUDA as the backend</span>

<span class="co"># Example RNA-seq dataset wrapped in an S4 class.</span>
<span class="kw">data</span>(paschold) 

<span class="co"># Use a small subset of the data (few genes).</span>
paschold@counts =<span class="st"> </span><span class="kw">head</span>(paschold@counts, <span class="dv">20</span>) 

<span class="co"># Run the MCMC for only a few iterations.</span>
configs =<span class="st"> </span><span class="kw">Configs</span>(<span class="dt">burnin =</span> <span class="dv">10</span>, <span class="dt">iterations =</span> <span class="dv">10</span>, <span class="dt">thin =</span> <span class="dv">1</span>) 

<span class="co"># Set up the MCMC.</span>
chain =<span class="st"> </span><span class="kw">Chain</span>(paschold, configs) 

<span class="co"># Run 4 dispersed independent Markov chains.</span>
chain_list =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">backend =</span> back_end)

<span class="co"># Monitor convergence with in all parameters with </span>
<span class="co"># Gelman-Rubin potential scale reduction factors</span>
<span class="kw">head</span>(<span class="kw">psrf</span>(chain_list)) 

<span class="co"># Means, standard deviations, and credible intervals of all parameters </span>
<span class="kw">head</span>(<span class="kw">estimates</span>(chain_list))

<span class="co"># Gene-specific (row-specific) posterior probabilities</span>
<span class="kw">head</span>(<span class="kw">probs</span>(chain_list))

<span class="co"># Monte Carlo samples on a small subset of parameters</span>
m =<span class="st"> </span><span class="kw">mcmc_samples</span>(chain_list) 
<span class="kw">dim</span>(m)
m[<span class="dv">1</span>:<span class="dv">5</span>, <span class="dv">1</span>:<span class="dv">5</span>]

<span class="co"># Set max_iter=Inf below to automatically run until convergence.</span>
max_iter =<span class="st"> </span><span class="dv">3</span>
iter =<span class="st"> </span><span class="dv">1</span>
chain@verbose =<span class="st"> </span><span class="kw">as.integer</span>(<span class="dv">0</span>) <span class="co"># turn off console messages</span>
chain_list =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">backend =</span> back_end)
<span class="co"># saveRDS(chain_list, &quot;chain_list_so_far.rds&quot;) # option to save progress</span>
gelman =<span class="st"> </span><span class="kw">psrf</span>(chain_list)
if(<span class="kw">any</span>(gelman &gt;=<span class="st"> </span><span class="fl">1.1</span>)) while(iter &lt;<span class="st"> </span>max_iter){
  chain_list =<span class="st"> </span><span class="kw">lapply</span>(chain_list, fbseq, <span class="dt">additional_chains =</span> <span class="dv">0</span>, <span class="dt">backend =</span> back_end)
  gelman =<span class="st"> </span><span class="kw">psrf</span>(chain_list)
  <span class="co"># saveRDS(chain_list, &quot;chain_list_so_far.rds&quot;) # option to save progress</span>
  if(<span class="kw">all</span>(gelman &lt;<span class="st"> </span><span class="fl">1.1</span>)) break
  iter =<span class="st"> </span>iter +<span class="st"> </span><span class="dv">1</span>
}</code></pre>
</div>
<div id="operational-details" class="section level1">
<h1><span class="header-section-number">7</span> Operational details</h1>
<p>Here is the breakdown for finer control of the package.</p>
<div id="create-a-scenario-object." class="section level2">
<h2><span class="header-section-number">7.1</span> Create a <code>Scenario</code> object.</h2>
<p><code>Scenario</code> is an S4 class for storing the count data, design matrix, and instructions for calculating posterior probabilities of quantities of interest.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">new</span>(<span class="st">&quot;Scenario&quot;</span>))</code></pre>
<pre><code>## Formal class 'Scenario' [package &quot;fbseq&quot;] with 6 slots
##   ..@ bounds      : num(0) 
##   ..@ contrasts   : list()
##   ..@ counts      : num[0 , 0 ] 
##   ..@ design      : num[0 , 0 ] 
##   ..@ propositions: list()
##   ..@ supplement  : list()</code></pre>
<p>Type <code>help(&quot;Scenario-class&quot;)</code> for details about the slots. When you have these slots ready, use the <code>Scenario</code> function to create a <code>Scenario</code> object.</p>
<pre class="sourceCode r"><code class="sourceCode r">scenario =<span class="st"> </span><span class="kw">Scenario</span>(<span class="dt">counts =</span> counts, <span class="dt">design =</span> design, 
  <span class="dt">contrasts =</span> contrasts, <span class="dt">bounds =</span> bounds, <span class="dt">propositions =</span> propositions, 
  <span class="dt">supplement =</span> supplement)</code></pre>
<p>For this constructor, arguments should be explicitly named. For example,</p>
<pre class="sourceCode r"><code class="sourceCode r">scenario =<span class="st"> </span><span class="kw">Scenario</span>(counts, design, contrasts, bounds, propositions, supplement)</code></pre>
<p>would throw an error.</p>
<div id="example-scenario-from-paschold" class="section level3">
<h3><span class="header-section-number">7.1.1</span> Example scenario from <span class="citation">Paschold et al. (2012)</span></h3>
<p>The package comes with a <code>Scenario</code> with data taken from <span class="citation">Paschold et al. (2012)</span>. The data is from an RNA-sequencing (RNA-seq) data experiment on maize.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(paschold)
<span class="kw">head</span>(paschold@counts)</code></pre>
<pre><code>##                  B73_1 B73_2 B73_3 B73_4 Mo17_1 Mo17_2 Mo17_3 Mo17_4
## AC147602.5_FG004     1     0     0     2      0      1      0      0
## AC148152.3_FG001     3     4     6     0      8     17     18     20
## AC148152.3_FG005  2323  1533  1932  1945   2070   1582   2196   1882
## AC148152.3_FG006     2     1     0     2      4      4      5      3
## AC148152.3_FG008     3     3     4     1     31     40     45     49
## AC148167.6_FG001   672   598   728   713    743    655    821    824
##                  B73xMo17_1 B73xMo17_2 B73xMo17_3 B73xMo17_4 Mo17xB73_1
## AC147602.5_FG004          1          1          1          0          0
## AC148152.3_FG001          7          8         19         16          7
## AC148152.3_FG005       1532       1271       2168       2004       1305
## AC148152.3_FG006          4          1          1          0          4
## AC148152.3_FG008          8          9          4          3          2
## AC148167.6_FG001        728        731        809        764        754
##                  Mo17xB73_2 Mo17xB73_3 Mo17xB73_4
## AC147602.5_FG004          0          0          0
## AC148152.3_FG001         10         13         13
## AC148152.3_FG005       1454       1694       1362
## AC148152.3_FG006          1          1          2
## AC148152.3_FG008          3          6          2
## AC148167.6_FG001        609        801        769</code></pre>
<p>The dataset is large, so if you’re just trying out the package on your home computer, it would be best to simply use a subset of the data</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(paschold@counts)</code></pre>
<pre><code>## [1] 39656    16</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">paschold@counts =<span class="st"> </span><span class="kw">head</span>(paschold@counts, <span class="dv">20</span>)
<span class="kw">dim</span>(paschold@counts)</code></pre>
<pre><code>## [1] 20 16</code></pre>
<p>In the data, the genes are rows and the columns are RNA-seq replicates. Each count is the relative measure of the expression of a gene in a replicate. The observations are divided into 4 genetic varieties:</p>
<ul>
<li><code>B73</code>: an inbred Iowa maize variety</li>
<li><code>Mo17</code>: an inbred Missouri maize variety</li>
<li><code>B73xMo17</code>: a first-generation (F1) hybrid produced from pollenating <code>B73</code> with <code>Mo17</code></li>
<li><code>Mo17xB73</code>: an F1 hybrid from pollenating <code>Mo17</code> with <code>B73</code></li>
</ul>
<p>The goal of this scenario is to detect genes that display heterosis, or hybrid vigor, with respect to expression level (RNA-seq count). Heterosis is the phenomenon in which a hybrid surpasses both parents with respect to a given trait. The different types of heterosis are defined as follows in terms of the log RNA-seq counts.</p>
<table>
<tbody>
<tr class="odd">
<td align="left">high-parent heterosis of the hybrid mean</td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left">hybrid mean &gt; max(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of the hybrid mean</td>
<td align="left"></td>
<td align="left">hybrid mean &lt; min(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis of <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><code>B73xMo17</code> &gt; max(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><code>B73xMo17</code> &lt; min(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis of <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><code>Mo17xB73</code> &gt; max(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><code>Mo17xB73</code> &lt; min(<code>B73</code> and <code>Mo17</code>)</td>
</tr>
</tbody>
</table>
<p>To define the parameterization of the model coefficients <span class="math">\(\beta_{\ell, g}\)</span>, we use the design matrix slot in our <code>Scenario</code> object.</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@design</code></pre>
<pre><code>##            beta_1 beta_2 beta_3 beta_4 beta_5
## B73_1           1      1     -1      0      1
## B73_2           1      1     -1      0      1
## B73_3           1      1     -1      0     -1
## B73_4           1      1     -1      0     -1
## Mo17_1          1     -1      1      0      1
## Mo17_2          1     -1      1      0      1
## Mo17_3          1     -1      1      0     -1
## Mo17_4          1     -1      1      0     -1
## B73xMo17_1      1      1      1      1      1
## B73xMo17_2      1      1      1      1      1
## B73xMo17_3      1      1      1      1     -1
## B73xMo17_4      1      1      1      1     -1
## Mo17xB73_1      1      1      1     -1      1
## Mo17xB73_2      1      1      1     -1      1
## Mo17xB73_3      1      1      1     -1     -1
## Mo17xB73_4      1      1      1     -1     -1</code></pre>
<p>We can interpret each parameters as follows.</p>
<table>
<tbody>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{1, g}\)</span></td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left">mean of <code>B73</code> and <code>Mo17</code></td>
</tr>
<tr class="even">
<td align="left"><span class="math">\(\beta_{2, g}\)</span></td>
<td align="left"></td>
<td align="left">half the difference between <code>Mo17</code> and the hybrid mean</td>
</tr>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{3, g}\)</span></td>
<td align="left"></td>
<td align="left">half the difference between <code>B73</code> and the hybrid mean</td>
</tr>
<tr class="even">
<td align="left"><span class="math">\(\beta_{4, g}\)</span></td>
<td align="left"></td>
<td align="left">difference between <code>B73xMo17</code> and <code>Mo17xB73</code></td>
</tr>
<tr class="odd">
<td align="left"><span class="math">\(\beta_{5, g}\)</span></td>
<td align="left"></td>
<td align="left">experimental block effect</td>
</tr>
</tbody>
</table>
<p>In terms of the <span class="math">\(\beta_{\ell, g}\)</span>’s, the conditions for heterosis become the following.</p>
<table>
<tbody>
<tr class="odd">
<td align="left">high-parent heterosis of the hybrid mean</td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left"><span class="math">\(\beta_{2, g} &gt; 0\)</span> and <span class="math">\(\beta_{3, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of the hybrid mean</td>
<td align="left"></td>
<td align="left"><span class="math">\(-\beta_{2, g} &gt; 0\)</span> and <span class="math">\(-\beta_{3, g} &gt; 0\)</span></td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis of <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(2\beta_{2, g} + \beta_{4, g} &gt; 0\)</span> and <span class="math">\(2\beta_{3, g} + \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of <code>B73xMo17</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(-2\beta_{2, g} - \beta_{4, g} &gt; 0\)</span> and <span class="math">\(-2\beta_{3, g} - \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="odd">
<td align="left">high-parent heterosis of <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(2\beta_{2, g} - \beta_{4, g} &gt; 0\)</span> and <span class="math">\(2\beta_{3, g} - \beta_{4, g} &gt; 0\)</span></td>
</tr>
<tr class="even">
<td align="left">low-parent heterosis of <code>Mo17xB73</code></td>
<td align="left"></td>
<td align="left"><span class="math">\(-2\beta_{2, g} + \beta_{4, g} &gt; 0\)</span> and <span class="math">\(-2\beta_{3, g} + \beta_{4, g} &gt; 0\)</span></td>
</tr>
</tbody>
</table>
<p>Gene-specific (i.e., row-specific) posterior probabilities are calculated using Monte Carlo samples <span class="math">\(\beta_{\ell, g}^{(m)}\)</span> of the <span class="math">\(\beta_{\ell, g}\)</span>’s for MCMC iteration <span class="math">\(m = 1, \ldots, M\)</span>. Below, <span class="math">\(\text{I}(\cdot)\)</span> is the indicator function (which returns 1 when its argument is true and 0 otherwise).</p>
<p><span class="math">\[ \begin{align*}
P(\text{high-parent heterosis, mean of hybrids} \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left (\beta_{2, g}^{(m)} &gt; 0 \text{ and } \beta_{3, g}^{(m)} &gt; 0 \right ) \\
P(\text{low-parent heterosis, mean of hybrids}) \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left ( -\beta_{2, g}^{(m)} &gt; 0 \text{ and } -\beta_{3, g}^{(m)} &gt; 0 \right ) \\
P(\text{high-parent heterosis, B73xMo17}) \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left ( 2\beta_{2, g}^{(m)} + \beta_{4, g}^{(m)} &gt; 0 \text{ and } 2\beta_{3, g}^{(m)} + \beta_{4, g}^{(m)} &gt; 0 \right ) \\
P(\text{low-parent heterosis, B73xMo17}) \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left ( -2\beta_{2, g}^{(m)} - \beta_{4, g}^{(m)} &gt; 0 \text{ and } -2\beta_{3, g}^{(m)} - \beta_{4, g}^{(m)} &gt; 0 \right ) \\
P(\text{high-parent heterosis, Mo17xB73}) \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left ( 2\beta_{2, g}^{(m)} - \beta_{4, g}^{(m)} &gt; 0 \text{ and } 2\beta_{3, g}^{(m)} - \beta_{4, g}^{(m)} &gt; 0 \right ) \\
P(\text{low-parent heterosis, Mo17xB73}) \ | \ \text{data}) &amp;\approx \frac{1}{M} \sum_{m = 1}^M \text{I}\left ( -2\beta_{2, g}^{(m)} + \beta_{4, g}^{(m)} &gt; 0 \text{ and }-2\beta_{3, g}^{(m)} + \beta_{4, g}^{(m)} &gt; 0 \right )
\end{align*} \]</span></p>
<p>To tell the software to calculate these posterior probabilities, use of the <code>contrasts</code>, <code>bounds</code>, and <code>propositions</code> slots of the <code>Scenario</code> object. In general, the package can calculate probabilities of that depend on linear combinations of the <span class="math">\(\beta_{\ell, g}\)</span>’s, and the <code>contrasts</code> slot specifies these linear combinations.</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@contrasts</code></pre>
<pre><code>## $`high-parent_B73xMo17_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      1      0      0      0 
## 
## $`high-parent_B73xMo17_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      1      0      0 
## 
## $`low-parent_B73xMo17_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -1      0      0      0 
## 
## $`low-parent_B73xMo17_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -1      0      0 
## 
## $`high-parent_B73xMo17_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0      1      0 
## 
## $`high-parent_B73xMo17_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2      1      0 
## 
## $`low-parent_B73xMo17_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0     -1      0 
## 
## $`low-parent_B73xMo17_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2     -1      0 
## 
## $`high-parent_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0     -1      0 
## 
## $`high-parent_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2     -1      0 
## 
## $`low-parent_Mo17xB73_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0      1      0 
## 
## $`low-parent_Mo17xB73_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2      1      0</code></pre>
<p>As in the heterosis example, the probabilities are calculated by comparing the linear combinations to lower bounds. In the heterosis example, linear combinations are being compared to zero, so all the slots in the <code>bounds</code> slot should be zero.</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@bounds</code></pre>
<pre><code>## high-parent_B73xMo17_Mo17xB73_1 high-parent_B73xMo17_Mo17xB73_2 
##                               0                               0 
##  low-parent_B73xMo17_Mo17xB73_1  low-parent_B73xMo17_Mo17xB73_2 
##                               0                               0 
##          high-parent_B73xMo17_1          high-parent_B73xMo17_2 
##                               0                               0 
##           low-parent_B73xMo17_1           low-parent_B73xMo17_2 
##                               0                               0 
##          high-parent_Mo17xB73_1          high-parent_Mo17xB73_2 
##                               0                               0 
##           low-parent_Mo17xB73_1           low-parent_Mo17xB73_2 
##                               0                               0</code></pre>
<p>The conditions for heterosis are formulated as logical propositions, and these propositions are logical conjunctions that depend on the linear combinations and lower bounds from before. To write these heterosis propositions in terms of <code>contrasts</code> and <code>bounds</code>, use the <code>propositions</code> slot.</p>
<pre class="sourceCode r"><code class="sourceCode r">paschold@propositions</code></pre>
<pre><code>## $`high-parent_B73xMo17_Mo17xB73`
## high-parent_B73xMo17_Mo17xB73_1 high-parent_B73xMo17_Mo17xB73_2 
##                               1                               2 
## 
## $`low-parent_B73xMo17_Mo17xB73`
## low-parent_B73xMo17_Mo17xB73_1 low-parent_B73xMo17_Mo17xB73_2 
##                              3                              4 
## 
## $`high-parent_B73xMo17`
## high-parent_B73xMo17_1 high-parent_B73xMo17_2 
##                      5                      6 
## 
## $`low-parent_B73xMo17`
## low-parent_B73xMo17_1 low-parent_B73xMo17_2 
##                     7                     8 
## 
## $`high-parent_Mo17xB73`
## high-parent_Mo17xB73_1 high-parent_Mo17xB73_2 
##                      9                     10 
## 
## $`low-parent_Mo17xB73`
## low-parent_Mo17xB73_1 low-parent_Mo17xB73_2 
##                    11                    12</code></pre>
<p>With the <code>contrasts</code>, <code>bounds</code>, and <code>propositions</code> defined, the-specific posterior probabilities are fully-specified.</p>
<p>Lastly, every <code>Scenario</code> object has an optional <code>suppelement</code> slot with additional information about the scenario. In this case, <code>paschold@supplement</code> is an empty list.</p>
</div>
<div id="example-simulated-scenario-with-scenario_heterosis_model" class="section level3">
<h3><span class="header-section-number">7.1.2</span> Example simulated scenario with <code>scenario_heterosis_model</code></h3>
<p>To simulate a heterosis scenario similar to that of <span class="citation">Paschold et al. (2012)</span>, use the <code>scenario_heterosis_model</code> function. Count data are simulated from the model given fixed values for the hyperparameters.</p>
<pre class="sourceCode r"><code class="sourceCode r">s =<span class="st"> </span><span class="kw">scenario_heterosis_model</span>(<span class="dt">genes =</span> <span class="dv">1000</span>, <span class="dt">libraries =</span> <span class="dv">16</span>)
<span class="kw">head</span>(s@counts)</code></pre>
<pre><code>##        parent1_1 parent1_2 parent1_3 parent1_4 parent2_5 parent2_6
## gene_1         4         8         3         7         6         7
## gene_2         3         7        10        11         5         3
## gene_3        12        12        37        46        33        30
## gene_4         2         3         1         1         3         5
## gene_5       188       141        94        87       135       154
## gene_6        12         5        10        15        17        16
##        parent2_7 parent2_8 hybrid1_9 hybrid1_10 hybrid1_11 hybrid1_12
## gene_1         7         5         1          3          2          4
## gene_2         8        13         6          3         16         12
## gene_3       131        93        24         17         74         75
## gene_4        10         3         5          6          5          6
## gene_5        73        75       148        144         78         89
## gene_6        11        13        21         16         19         20
##        hybrid2_13 hybrid2_14 hybrid2_15 hybrid2_16
## gene_1          4          4          7          1
## gene_2          9          9          6          9
## gene_3         30         20         74         72
## gene_4          8          9          9          5
## gene_5        139        122         77         77
## gene_6         14         18         12         10</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">s@design</code></pre>
<pre><code>##            beta_1 beta_2 beta_3 beta_4 beta_5
## parent1_1       1      1     -1      0      1
## parent1_2       1      1     -1      0      1
## parent1_3       1      1     -1      0     -1
## parent1_4       1      1     -1      0     -1
## parent2_5       1     -1      1      0      1
## parent2_6       1     -1      1      0      1
## parent2_7       1     -1      1      0     -1
## parent2_8       1     -1      1      0     -1
## hybrid1_9       1      1      1      1      1
## hybrid1_10      1      1      1      1      1
## hybrid1_11      1      1      1      1     -1
## hybrid1_12      1      1      1      1     -1
## hybrid2_13      1      1      1     -1      1
## hybrid2_14      1      1      1     -1      1
## hybrid2_15      1      1      1     -1     -1
## hybrid2_16      1      1      1     -1     -1</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">s@contrasts</code></pre>
<pre><code>## $`high-parent_hybrids_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      1      0      0      0 
## 
## $`high-parent_hybrids_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      1      0      0 
## 
## $`low-parent_hybrids_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -1      0      0      0 
## 
## $`low-parent_hybrids_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -1      0      0 
## 
## $`high-parent_hybrid1_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0      1      0 
## 
## $`high-parent_hybrid1_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2      1      0 
## 
## $`low-parent_hybrid1_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0     -1      0 
## 
## $`low-parent_hybrid1_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2     -1      0 
## 
## $`high-parent_hybrid2_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      2      0     -1      0 
## 
## $`high-parent_hybrid2_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0      2     -1      0 
## 
## $`low-parent_hybrid2_1`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0     -2      0      1      0 
## 
## $`low-parent_hybrid2_2`
## beta_1 beta_2 beta_3 beta_4 beta_5 
##      0      0     -2      1      0</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">s@bounds</code></pre>
<pre><code>## high-parent_hybrids_1 high-parent_hybrids_2  low-parent_hybrids_1 
##                     0                     0                     0 
##  low-parent_hybrids_2 high-parent_hybrid1_1 high-parent_hybrid1_2 
##                     0                     0                     0 
##  low-parent_hybrid1_1  low-parent_hybrid1_2 high-parent_hybrid2_1 
##                     0                     0                     0 
## high-parent_hybrid2_2  low-parent_hybrid2_1  low-parent_hybrid2_2 
##                     0                     0                     0</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">s@propositions</code></pre>
<pre><code>## $`high-parent_hybrids`
## high-parent_hybrids_1 high-parent_hybrids_2 
##                     1                     2 
## 
## $`low-parent_hybrids`
## low-parent_hybrids_1 low-parent_hybrids_2 
##                    3                    4 
## 
## $`high-parent_hybrid1`
## high-parent_hybrid1_1 high-parent_hybrid1_2 
##                     5                     6 
## 
## $`low-parent_hybrid1`
## low-parent_hybrid1_1 low-parent_hybrid1_2 
##                    7                    8 
## 
## $`high-parent_hybrid2`
## high-parent_hybrid2_1 high-parent_hybrid2_2 
##                     9                    10 
## 
## $`low-parent_hybrid2`
## low-parent_hybrid2_1 low-parent_hybrid2_2 
##                   11                   12</code></pre>
</div>
</div>
<div id="optional-create-a-configs-object." class="section level2">
<h2><span class="header-section-number">7.2</span> (Optional) create a <code>Configs</code> object.</h2>
<p>The <code>Configs</code> S4 class contains the practical MCMC configuration parameters, such as the number of Monte Carlo iterations and the length of burnin, and important details about model specification. All of them have default settings, so just for the sake of trying out the package, you should be able to get away with calling the <code>Configs</code> function with no arguments. For a full list of slots and their explanations, type <code>help(&quot;Configs-class&quot;)</code>.</p>
<p>The main slots to consider are the <code>iterations</code>, <code>burnin</code>, and <code>thin</code> slots, which control the duration of the MCMC. The total number of MCMC iterations is <code>burnin + iterations * thin</code>. <code>burinin</code> is the number of throwaway iterations at the beginning of the MCMC. Afterwards the next <code>iterations * thin</code> iterations are used to compute posterior probabilities and estimated means and mean squares for every parameter. Only <code>iterations</code> Monte Carlo parameter samples are returned to the user, and then, only for parameters specified in the <code>parameter_sets_return</code> slot. The <code>parameter_sets_update</code> slot specifies which parameters to update during the MCMC, giving the user the option to run empirical Bayes, for example. For gene-specific parameters in <code>parameter_sets_update</code>, not all parameters are returned for all the genes. (Downloading data from the GPU is expensive). Select the genes for parameter samples you want with the <code>genes_return</code> slot. For the <span class="math">\(\epsilon_{n, g}\)</span> parameters, use <code>genes_return_epsilon</code> to specify the values of <span class="math">\(g\)</span> and <code>libraries_return_epsilon</code> to specify the values of <span class="math">\(n\)</span>.</p>
</div>
<div id="optional-create-a-starts-object." class="section level2">
<h2><span class="header-section-number">7.3</span> (Optional) create a <code>Starts</code> object.</h2>
<p>The <code>Starts</code> S4 class has the starting values of an MCMC chain. Usually, the user does not need to worry about it, as the unspecified starting values are automatically generated. However, the user may wish to specify starting values by setting the slots manually. For more information, type <code>help(&quot;Starts-class&quot;)</code> and be sure to read the vignette on the hierarchical model.</p>
</div>
<div id="create-a-chain-object." class="section level2">
<h2><span class="header-section-number">7.4</span> Create a <code>Chain</code> object.</h2>
<p>The <code>Chain</code> class is the container for a whole Markov chain. It stores all the information in your <code>Scenario</code>, <code>Configs</code>, and <code>Starts</code> objects, in addition to Monte Carlo output and results. There are many, many S4 slots. For details on the slots, type <code>help(&quot;Chain-class&quot;)</code> and be sure to read the hierarchical model vignette. For help on how to create <code>Chain</code> objects, type <code>help(&quot;Chain&quot;)</code>.</p>
<p>With only a <code>Scenario</code> object, you can create a <code>Chain</code> object with</p>
<pre class="sourceCode r"><code class="sourceCode r">chain =<span class="st"> </span><span class="kw">Chain</span>(scenario)</code></pre>
<p>If you have <code>Configs</code> and/or <code>Starts</code> objects, you can pass them in as well.</p>
<pre class="sourceCode r"><code class="sourceCode r">chain =<span class="st"> </span><span class="kw">Chain</span>(scenario, configs, starts)</code></pre>
<p>As an example, here is the <code>Chain</code> created from the <code>paschold</code> scenario visited earlier.</p>
<pre class="sourceCode r"><code class="sourceCode r">chain =<span class="st"> </span><span class="kw">Chain</span>(paschold)
<span class="co"># Call str(chain) to inspect the slots.</span></code></pre>
<p>You can recover the scenario, configuration, or starting values of a <code>Chain</code> object with the <code>Scenario</code>, <code>Configs</code>, and <code>Starts</code> objects. For example,</p>
<pre class="sourceCode r"><code class="sourceCode r">scenario =<span class="st"> </span><span class="kw">Scenario</span>(chain)
configs =<span class="st"> </span><span class="kw">Configs</span>(chain)
starts =<span class="st"> </span><span class="kw">Starts</span>(chain)
<span class="kw">str</span>(configs)</code></pre>
<pre><code>## Formal class 'Configs' [package &quot;fbseq&quot;] with 13 slots
##   ..@ burnin                  : int 100000
##   ..@ effects_update_beta     : int [1:5] 1 2 3 4 5
##   ..@ genes_return            : int [1:3] 13602 25867 31079
##   ..@ genes_return_epsilon    : int [1:3] 11899 31199 39117
##   ..@ iterations              : int 5000
##   ..@ libraries_return        : int [1:3] 1 2 14
##   ..@ libraries_return_epsilon: int [1:3] 1 7 16
##   ..@ parameter_sets_return   : chr [1:7] &quot;beta&quot; &quot;epsilon&quot; &quot;gamma&quot; &quot;nu&quot; ...
##   ..@ parameter_sets_update   : chr [1:7] &quot;beta&quot; &quot;epsilon&quot; &quot;gamma&quot; &quot;nu&quot; ...
##   ..@ priors                  : Named chr [1:5] &quot;normal&quot; &quot;normal&quot; &quot;normal&quot; &quot;normal&quot; ...
##   .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;normal&quot; &quot;normal&quot; &quot;normal&quot; &quot;normal&quot; ...
##   ..@ samplers                : chr &quot;default&quot;
##   ..@ thin                    : num 20
##   ..@ verbose                 : int 5</code></pre>
<p>Note that the unspecified starting values were automatically generated in the construction of the <code>Chain</code> object.</p>
</div>
<div id="call-fbseq-on-the-chain-object." class="section level2">
<h2><span class="header-section-number">7.5</span> Call <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> on the <code>Chain</code> object.</h2>
<p>The <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> function runs the MCMC algorithm to estimate parameters and calculate posterior probabilities. It takes a single <code>Chain</code> object as an argument, and outputs either another <code>Chain</code> object or a list of <code>Chain</code> objects, depending on the value of <code>additional_chains</code>. To run a single chain, use:</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">additional_chains =</span> <span class="dv">0</span>)</code></pre>
<p>Here, <code>out</code> is another <code>Chain</code> object with the results. The starting values of <code>out</code> are the final parameter samples of the previous MCMC. That way, you can call <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> on <code>out</code> and the MCMC will resume where it last left off.</p>
<pre class="sourceCode r"><code class="sourceCode r">continuation =<span class="st"> </span><span class="kw">fbseq</span>(out, <span class="dt">additional_chains =</span> <span class="dv">0</span>)</code></pre>
<p>To monitor convergence, you can run multiple chains and use the <code>psrf</code> function to compute Gelman-Rubin potential scale reduction factors on a list of chains <span class="citation">(Gelman et al. 2013)</span>. For example, to automatically run the MCMC until convergence, you could do the following.</p>
<pre class="sourceCode r"><code class="sourceCode r">chain_list =<span class="st"> </span><span class="kw">fbseq</span>(chain)  <span class="co"># equivalent to fbseq(chain, additional_chains = 3)</span>
<span class="kw">saveRDS</span>(chain, <span class="st">&quot;chain_list_so_far.rds&quot;</span>)
if(<span class="kw">any</span>(<span class="kw">psrf</span>(chain_list)) &gt;=<span class="st"> </span><span class="fl">1.1</span>) repeat{
  chain_list =<span class="st"> </span><span class="kw">lapply</span>(chain_list, fbseq)
  <span class="kw">saveRDS</span>(chain, <span class="st">&quot;chain_list_so_far.rds&quot;</span>)
  if(<span class="kw">all</span>(<span class="kw">psrf</span>(chain_list)) &lt;<span class="st"> </span><span class="fl">1.1</span>) break
}</code></pre>
<p>If the <code>additional_chains</code> argument is greater than 0, an additional <code>additional_chains</code> MCMC chains will be run after the first one, and the output object from <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> will a list of the total <code>1 + additional_chains</code> run. This is useful for computing Gelman-Rubin potential scale reduction factors with the <code>psrf</code> function to assess convergence.</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">additional_chains =</span> <span class="dv">3</span>)
<span class="kw">psrf</span>(out)</code></pre>
</div>
<div id="choosing-a-backend-fbseqserial-or-fbseqcuda" class="section level2">
<h2><span class="header-section-number">7.6</span> Choosing a backend: <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> or <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a></h2>
<p><a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> can run on most machines, but it is slow for datasets with lots of genes. Select the <a href="https://github.com/wlandau/fbseqSerial"><code>fbseqSerial</code></a> backend by specifying <code>backend = &quot;serial&quot;</code> in <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> :</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">backend =</span> <span class="st">&quot;serial&quot;</span>)</code></pre>
<p>For the serial backend, you can distribute the MCMC chains after the first chain over multiple parallel processes.</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">backend =</span> <span class="st">&quot;serial&quot;</span>, <span class="dt">processes =</span> <span class="dv">3</span>)</code></pre>
<p><a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> requires a CUDA-capable graphics processing unit (GPU), but it is much faster. Elapsed runtimes are in hours rather than days for typical-sized RNA-seq datasets. Select this backend with either</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain, <span class="dt">backend =</span> <span class="st">&quot;CUDA&quot;</span>)</code></pre>
<p>or simply</p>
<pre class="sourceCode r"><code class="sourceCode r">out =<span class="st"> </span><span class="kw">fbseq</span>(chain)</code></pre>
</div>
<div id="inspect-the-output." class="section level2">
<h2><span class="header-section-number">7.7</span> Inspect the output.</h2>
<p>The <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> package has several functions for analyzing MCMC output. Below <code>obj</code>, can either be a <code>Chain</code> object or a list of <code>Chain</code> objects.</p>
<table>
<tbody>
<tr class="odd">
<td align="left"><code>psrf(obj)</code></td>
<td align="left"></td>
<td align="left">Gelman-Rubin potential scale reduction factors on all parameters, used to monitor convergence <span class="citation">(Gelman et al. 2013)</span>. Here, <code>obj</code> must be a list of at least 2 <code>Chain</code> objects.</td>
</tr>
<tr class="even">
<td align="left"><code>probs(obj)</code></td>
<td align="left"></td>
<td align="left">Estimated gene-specific posterior probabilities specified in <code>Scenario</code></td>
</tr>
<tr class="odd">
<td align="left"><code>effect_sizes(obj)</code></td>
<td align="left"></td>
<td align="left">Effect sizes that correspond to the posterior probabilities. For example, for “<span class="math">\(P (2\beta_{2, g} - \beta_{4, g} &gt; 1\)</span> and <span class="math">\(2\beta_{3, g} - \beta_{4, g} &gt; 0.5 | \text{data})\)</span>”, the effect size is the positive part of <span class="math">\(\min \left (2\beta_{2, g} - \beta_{4, g} - 1, \ 2\beta_{3, g} - \beta_{4, g} - 0.5 \right )\)</span>.</td>
</tr>
<tr class="even">
<td align="left"><code>estimates(obj, level)</code></td>
<td align="left"></td>
<td align="left">Posterior means, standard deviations, and credible intervals (of level <code>level</code>) for all parameters. The credible intervals are computed using the normal distributions given by the posterior means and standard deviations.</td>
</tr>
<tr class="odd">
<td align="left"><code>mcmc_samples(obj)</code></td>
<td align="left"><span class="math">\(\qquad\)</span></td>
<td align="left">Monte Carlo samples of returned parameters selected in <code>Configs</code></td>
</tr>
</tbody>
</table>
<div class="references">
<h1 id="references" class="unnumbered">References</h1>
<p>Gelman, Andrew, John b. Carlin, Hal S. Stern, David B. Dunson, Aki Vehtari, and Donald B. Rubin. 2013. <em>Bayesian Data Analysis</em>. 3rd ed. CRC Press.</p>
<p>Paschold, Anja, Yi Jia, Caroline Marcon, Steve Lund, Nick B Larson, Cheng-Ting Yeh, Stephan Ossowski, et al. 2012. “Complementation Contributes to Transcriptome Complexity in Maize (<em>Zea Mays L.</em>) Hybrids Relative to Their Inbred Parents.” <em>Genome Research</em> 22 (12). Cold Spring Harbor Lab: 2445–54.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
